modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Traverse.kt

```diff
@@ -5,11 +5,519 @@ import arrow.core.Id
+ *
+ * Moreover, the fact that our aggregate job is not tied to any specific `SparkContext` allows us to pass in a `SparkContext` pointing to a production cluster, or (using the exact same job) pass in a test `SparkContext` that just runs locally across threads. This makes testing our large job nice and easy.
+ *
+ * Finally, this encoding ensures that all the jobs for each topic run on the exact same cluster. At no point do we manually pass in or thread a `SparkContext` through - that is taken care for us by the (applicative) behavior of `Reader` and therefore by [traverse].
+ *
+ * ### Traversables are Functors
+ *
+ * As it turns out every [Traverse] is a lawful [Functor]. By carefully picking the `G` to use in [traverse] we can implement `map`.
+ *
+ * First let's look at the two signatures.
+ *
+ * ```kotlin:ank
+ * import arrow.Kind
+ * import arrow.core.Id
+ * import arrow.core.extensions.id.applicative.applicative
+ * import arrow.core.value
+ * import arrow.typeclasses.Applicative
+ * import arrow.typeclasses.Foldable
+ * import arrow.typeclasses.Functor
+ *
+ * interface Traverse<F> : Functor<F>, Foldable<F> {
+ *   fun <G, A, B> Kind<F, A>.traverse(AP: Applicative<G>, f: (A) -> Kind<G, B>): Kind<G, Kind<F, B>>
+ *
+ *   override fun <A, B> Kind<F, A>.map(f: (A) -> B): Kind<F, B> =
+ *     traverse(Id.applicative()) { Id(f(it)) }.value()
+ * }
+ * ```
+ *
+ * Both have an `Kind<F, A>` receiver and a similar `f` parameter. [traverse] expects the return type of `f` to be `Kind<G, B>` whereas `map` just wants `B`. Similarly the return type of [traverse] is `Kind<G, Kind<F, B>>` whereas for `map` it's just `Kind<F, B>`. This suggests we need to pick a `G` such that `Kind<G, A>` communicates exactly as much information as `A`. We can conjure one up by simply wrapping an `A` in `arrow.core.Id`.
+ *
+ * In order to call [traverse] [Id] needs to be [Applicative] which is straightforward - note that while [Id] just wraps an `A`, it is still a type constructor which matches the shape required by [Applicative].
+ *
+ * ```kotlin:ank
+ * import arrow.core.ForId
+ * import arrow.core.Id
+ * import arrow.core.IdOf
+ * import arrow.core.fix
+ * import arrow.typeclasses.Applicative
+ *
+ * interface IdApplicative : Applicative<ForId> {
+ *   override fun <A, B> IdOf<A>.ap(ff: IdOf<(A) -> B>): Id<B> =
+ *     ff.fix().flatMap { f -> map(f) }.fix()
```

Comment:
```suggestion
 *     Id(ff.value().invoke(this.value()))
```
I know that the flatMap one is the one used in Id, and it is correct, however it doesn't show what is actually happening. Either do the above or include the flatMap definition somewhere.